### ISSUE TITLE
Strange Bug with Macro Substitution?

### ISSUE THREAD

**Issue by DavePearce** [Whiley/WhileyTheoremProver#67]
There is a very strange problem with the following (from the resize example):

<pre>
define unchanged(int[] xs, int n, int[] ys) is:
    forall(int j):
        if:
            (0 <= j) && (j < |xs|)
        then:
            (j >= n) || (xs[j] == ys[j])

define resize_ensures_2(int[] xs, nat n, int v, int[] ys) is:
    forall(int i).(((|xs| <= i) && (i < n)) ==> (ys[i] == v))

define resize_loopinvariant_31(int[] xs, nat n, int[] rs, nat i) is:
    |rs| == n

define resize_loopinvariant_35(int[] xs, nat n, int v, int[] rs, nat i) is:
    unchanged(xs, i, rs)

define resize_loopinvariant_45(int[] xs, nat n, int v, int[] rs) is:
    forall(int j):
        if:
            ((|xs| <= j) && (j < n))
        then:
            rs[j] == v

assert "postcondition not satisfied":
    forall(int v, int[] xs, nat n, int[] rs, int i):
        if:
            resize_loopinvariant_31(xs, n, rs, i)
            resize_loopinvariant_35(xs, n, v, rs, i)
            resize_loopinvariant_45(xs, n, v, rs)
        then:
            resize_ensures_2(xs, n, v, rs)

type nat is (int x)
where:
    x >= 0
</pre>

Basically, if you do just about anything to this (e.g. inline one of the macros) then it will verify.  And. ss far as I'm aware, it is indeed correct.

**Comment by DavePearce** [comment]
Right, so when it fails it does so because of the classic "counting" problem.  This may be arising from the fact that it infers this (recursive) formula:
<pre>
rs[i] == rs[5 + rs[i]] 
</pre>

**Comment by DavePearce** [comment]
I devised this from the above, though it doesn't trigger loop yet:

<pre>
define invariant(int[] rs, int[] xs, int l) is:
    forall(int m):
        (rs[l] == rs[m]) || (m >= |rs|) || (|xs| >= (1 + m))
        
assert:
    forall(int[] rs, int[] xs, int l):
        if:
            invariant(rs,xs,l)
        then:
            |rs| < (1 + rs[l])
</pre>

**Comment by DavePearce** [comment]
Roughly speaking, quantifier instantiation occurs whenever the entire quantified formula contains a trigger.  However, I think we want to be a little more subjective than that?

**Comment by DavePearce** [comment]
Right, the problem is (as always) to do with representation:

<pre>
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ 460. rs[l] == xs[l]                                                                                       (Or-E 462) │
│┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
││ 464. v >= (1 + rs[l])                                                                                    (Or-E 467) │
││ 472. v >= (1 + xs[l])                                                                                (Eq-S 464,460) │
││ 476. (v == rs[l]) || (l >= |rs|) || (|xs| >= (1 + l))                                            (Forall-I 433,460) │
││┌────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│││ 473. v == rs[l]                                                                                         (Or-E 476) │
│││ 479. forall(int m).((rs[l] == rs[m]) || (m >= |rs|) || (|xs| >= (1 + m)))                           (Eq-S 433,473) │
│││ 480. rs[l] >= (1 + xs[l])                                                                           (Eq-S 472,473) │
│││ 482. true                                                                                       (Forall-I 479,473) │
</pre>

**Comment by DavePearce** [comment]
Therefore, this should be resolved by fixing the congruence interface (#63).
